### PR TITLE
Feat/1650 convert video to mp4 on upload

### LINKED ISSUE #1650: Convert video to mp4 on upload

### PR DESCRIPTION
**Pull request by frolyo**

### CODE REVIEW

File: src/dotnet/Core/Uploads/IUploadProcessor.cs [hunk 1]
```diff
@@ -1,11 +1,57 @@
+using Stl.IO;
+
 namespace ActualChat.Uploads;
 
 public interface IUploadProcessor
```
Comment (crui3er) [inline src/dotnet/Core/Uploads/IUploadProcessor.cs:3]: Split  types in independent files

File: src/dotnet/Core/Uploads/UploadProcessorsExt.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace ActualChat.Uploads;
+
+public static class UploadProcessorsExt
+{
+    public static async Task<ProcessedFile> Process(this IEnumerable<IUploadProcessor> processors, FilePath fileName, string contentType, Stream stream, CancellationToken cancellationToken)
+    {
+        var processor = processors.FirstOrDefault(x => x.Supports(contentType));
+        if (processor is null) {
+            // no need to dump file to file system
+            var memoryFile = await ToMemoryFile(fileName, contentType, stream, cancellationToken).ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/Core/Uploads/UploadProcessorsExt.cs]: Why do we copy file content to memory?
Can't we refer to original stream?

File: src/dotnet/Core/Uploads/VideoUploadProcessor.cs [hunk 3]
```diff
@@ -1,38 +1,71 @@
...
 
-    public async Task<ProcessedFileInfo> Process(FileInfo file, CancellationToken cancellationToken)
+    public async Task<ProcessedFile> Process(UploadedFile file, CancellationToken cancellationToken)
     {
-        var size = await GetVideoDimensions(file, cancellationToken).ConfigureAwait(false);
-        return size != null
-            ? new ProcessedFileInfo(file, size)
-            : new ProcessedFileInfo(file with { ContentType = MediaTypeNames.Application.Octet, }, null);
+        var (needConversion, size) = await GetVideoInfo(file, cancellationToken).ConfigureAwait(false);
+        if (size == null)
```
Comment (crui3er) [inline src/dotnet/Core/Uploads/VideoUploadProcessor.cs:19]: does size == null mean we failed to read video file info?

File: src/dotnet/Core/Uploads/VideoUploadProcessor.cs [hunk 4]
```diff
@@ -1,38 +1,71 @@
...
+            file = file with { ContentType = MediaTypeNames.Application.Octet };
```
Comment (crui3er) [inline src/dotnet/Core/Uploads/VideoUploadProcessor.cs]: Why do we replace ContentType here?

Comment (frolyo) [reply src/dotnet/Core/Uploads/VideoUploadProcessor.cs]: because we convert to mp4

File: src/dotnet/Core/Uploads/IUploadProcessor.cs (hunk 1 above)
Comment (frolyo) [reply src/dotnet/Core/Uploads/IUploadProcessor.cs:3]: done

File: src/dotnet/Core/Uploads/VideoUploadProcessor.cs (hunk 3 above)
Comment (frolyo) [reply src/dotnet/Core/Uploads/VideoUploadProcessor.cs:19]: right, or if video file has no video streams

File: src/dotnet/Core/Uploads/UploadProcessorsExt.cs (hunk 2 above)
Comment (frolyo) [reply src/dotnet/Core/Uploads/UploadProcessorsExt.cs]: done

**Pull request merged**